### PR TITLE
HHH-13060 - Throw an exception when two entities are declared with the same name

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-13060

### CODE REVIEW

File: documentation/src/main/asciidoc/userguide/chapters/domain/entity.adoc [hunk 1]
```diff
@@ -117,6 +117,13 @@ The main piece in mapping the entity is the `javax.persistence.Entity` annotatio
 The `@Entity` annotation defines just one attribute `name` which is used to give a specific entity name for use in JPQL queries.
 By default, the entity name represents the unqualified name of the entity class itself.
 
+[IMPORTANT]
+====
+Because the entity name is given by the unqualified name of the class, Hibernate does not allow registering multiple entities with the same name even if the entity classes reside in different packages.
+
+Without imposing this restriction, Hibernate would not know which entity class is referenced in a JPQL query if the unqualifid entity name is associated with more then one entity classes.
```
Comment (gsmet) [inline documentation/src/main/asciidoc/userguide/chapters/domain/entity.adoc]: s/unqualifid/unqualified/

Maybe we should say a word about specifying a different name in `@Entity`?

File: documentation/src/main/asciidoc/userguide/chapters/envers/Envers.adoc [hunk 2]
```diff
@@ -1028,7 +1028,12 @@ This basic query allows retrieving entity names and corresponding Java classes c
 ====
 [source, JAVA, indent=0]
 ----
-include::{sourcedir}/EntityTypeChangeAuditTest.java[tags=envers-tracking-modified-entities-queries-example]
+include::{sourcedir}/EntityTypeChangeAuditTest.java[tags=envers-tracking-modified-entities-queries-example1]
+----
+
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/EntityTypeChangeAuditTest.java[tags=envers-tracking-modified-entities-queries-example2]
```
Comment (gsmet) [inline documentation/src/main/asciidoc/userguide/chapters/envers/Envers.adoc]: How is it related to this issue?

This commit is quite hard to follow :/

File: hibernate-core/src/test/java/org/hibernate/test/annotations/access/Furniture.java [hunk 3]
```diff
@@ -19,38 +19,6 @@
  */
 @Entity
 @AccessType("field")
-public class Furniture extends Woody {
-	@Id
-	@GeneratedValue
-	private Integer id;
+public class Furniture extends BaseFurniture {
```
Comment (gsmet) [inline hibernate-core/src/test/java/org/hibernate/test/annotations/access/Furniture.java]: What does it have to do with this issue?

File: hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java [hunk 4]
```diff
@@ -31,13 +32,14 @@ public void testDuplicateEntityName() throws Exception {
 		try {
 			cfg.addAnnotatedClass( Flight.class );
 			cfg.addAnnotatedClass( org.hibernate.test.annotations.Flight.class );
+			cfg.addAnnotatedClass( org.hibernate.test.annotations.Company.class );
```
Comment (gsmet) [inline hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java]: Why add this?

File: hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java [hunk 5]
```diff
@@ -31,13 +32,14 @@ public void testDuplicateEntityName() throws Exception {
...
 			cfg.addResource( "org/hibernate/test/annotations/orm.xml" );
 			cfg.addResource( "org/hibernate/test/annotations/duplicatedgenerator/orm.xml" );
 			serviceRegistry = ServiceRegistryBuilder.buildServiceRegistry( cfg.getProperties() );
 			sf = cfg.buildSessionFactory( serviceRegistry );
             Assert.fail( "Should not be able to map the same entity name twice" );
 		}
-		catch (AnnotationException ae) {
+		catch (DuplicateMappingException ae) {
```
Comment (gsmet) [inline hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java]: So we had an exception before if the class names were conflicting? How comes the OP didn't have an error?

File: documentation/src/main/asciidoc/userguide/chapters/envers/Envers.adoc (hunk 2 above)
Comment (vladmihalcea) [reply documentation/src/main/asciidoc/userguide/chapters/envers/Envers.adoc]: The test fails now due to the change as previously we allowed entities having the same JPA name but with a different  class name.

File: hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java (hunk 5 above)
Comment (vladmihalcea) [reply hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java]: Nope. This test was actually failing because of a missing class. It was useless.

File: hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java (hunk 4 above)
Comment (vladmihalcea) [reply hibernate-core/src/test/java/org/hibernate/test/annotations/duplicatedgenerator/DuplicateTest.java]: Because the test was wrong. The Flight entity has an association to Company.

File: hibernate-core/src/test/java/org/hibernate/test/annotations/access/Furniture.java (hunk 3 above)
Comment (vladmihalcea) [reply hibernate-core/src/test/java/org/hibernate/test/annotations/access/Furniture.java]: Because we used to have an embeddable extend an entity which is wrong. This change fixes tgat as otherwise the embeddable would be register as an entity.

File: documentation/src/main/asciidoc/userguide/chapters/domain/entity.adoc (hunk 1 above)
Comment (vladmihalcea) [reply documentation/src/main/asciidoc/userguide/chapters/domain/entity.adoc]: I'll check that.

**Comment by vladmihalcea** [comment]
@gsmet Of course all changes are relevant to this change. Mostly because the tests were badly constructed.

I'll have to add a comment about the entity name provided explicitly.

**Comment by gsmet** [comment]
@vladmihalcea OK, it wasn't easy to understand from the various commits :).

Yes, please add a comment about the name and I think it's ready to go.
